internal/controller/reconciler/credentials.go

```diff
@@ -0,0 +1,100 @@
+package reconciler
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/controller/atlas"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/translation/project"
+)
+
+const (
+	orgIDKey      = "orgId"
+	publicAPIKey  = "publicApiKey"
+	privateAPIKey = "privateApiKey"
+)
+
+func (r *AtlasReconciler) ResolveCredentials(ctx context.Context, referrer project.ProjectReferrerObject) (*atlas.Credentials, error) {
+	connectionSecret := r.connectionSecretRef(referrer)
+	if connectionSecret != nil && connectionSecret.Name != "" {
+		creds, err := GetAtlasCredentials(ctx, r.Client, connectionSecret, &r.GlobalSecretRef)
+		if err != nil {
+			return nil, fmt.Errorf("error getting credentials from connection secret: %w", err)
+		}
+		return creds, nil
+	}
+
+	prj, err := r.fetchProject(ctx, referrer)
+	if err != nil {
+		return nil, fmt.Errorf("error resolving project reference: %w", err)
+	}
+
+	var projectSecret *client.ObjectKey
+	if prj != nil {
+		projectSecret = prj.ConnectionSecretObjectKey()
+	}
+
+	creds, err := GetAtlasCredentials(ctx, r.Client, projectSecret, &r.GlobalSecretRef)
+	if err != nil {
+		return nil, fmt.Errorf("error getting credentials from project reference: %w", err)
+	}
+	return creds, nil
+}
+
+func (r *AtlasReconciler) connectionSecretRef(pro project.ProjectReferrerObject) *client.ObjectKey {
+	key := client.ObjectKeyFromObject(pro)
+	pdr := pro.ProjectDualRef()
+	if pdr.ConnectionSecret == nil {
+		return nil
+	}
+	key.Name = pdr.ConnectionSecret.Name
+	return &key
+}
+
+func GetAtlasCredentials(ctx context.Context, k8sClient client.Client, secretRef, fallbackRef *client.ObjectKey) (*atlas.Credentials, error) {
```

Comment:
Given my suggestion on the provider, I'd return the org id separate from credentials.
```suggestion
func GetAtlasCredentials(ctx context.Context, k8sClient client.Client, secretRef, fallbackRef *client.ObjectKey) (string, *atlas.Credentials, error) {
```

Reply:
No, the orgId is specifically part of the public/private key tuple, this might be different for other types of credentials, i.e. OAuth opaque access tokens where we have no information about the encoded orgID.